### ISSUE TITLE
Add default validators for common formats date and datetime

### ISSUE THREAD

**Issue by bnchrch** [airbytehq/airbyte#24128]

**Comment by flash1293** [comment]
Back when the date picker got added to the UI, I asked the same question: https://github.com/airbytehq/airbyte/pull/19678#issuecomment-1337728373

This was @josephkmh response back then:
> I'm not sure I have enough context on the [date types](https://docs.airbyte.com/understanding-airbyte/supported-data-types/#dates-and-timestamps) (there are five) to know if this would break anything - probably not? But my feeling is that keeping the connector definition specification as the source of truth for validation is the best approach. I'd rather advocate for consistency in the connector specs than patching missing fields on the frontend.

@bnchrch @erohmensing you have a better feeling for this than me, is this still an issue?

**Comment by erohmensing** [comment]
I _think_ that link isn't relevant as it pertains to connector catalogs/configured catalogs and not specs/configs. The same stuff around which ben was asking about `airbyte_type`s. I think that `airbyte_type`s were introduced to solve issues like this, and that documentation might be out of date now 🤔

**Comment by erohmensing** [comment]
Actually further down that is mentioned: 
> The airbyte_type field resolves this ambiguity; sources producing timestamp-ish fields should choose either timestamp_with_timezone or timestamp_without_timezone (or time with/without timezone).

But yeah, I don't think that applies to specs here

**Comment by flash1293** [comment]
Summarizing the discussions:

This is the current state:
* Acceptance tests warn if the format and the pattern doesn't match (but it doesn't fail validation)
* The UI treats the pattern as blocking (if the input doesn't match the pattern the form can't be submitted)
* The UI treats the format as a hint of the field semantics - if it's set, the date(time) picker is rendered, but it doesn't enforce the format (it's still possible to type arbitrary strings into the input)


Desired state:

The `format` field property in JSON schema is not just a hint for field semantics, it's also a "sub type". Quoting the specification here (https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-validation-01#section-7)

```
The "format" keyword functions as both an annotation ([Section 3.3](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-validation-01#section-3.3))
   and as an assertion ([Section 3.2](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-validation-01#section-3.2)).  While no special effort is
   required to implement it as an annotation conveying semantic meaning,
   implementing validation is non-trivial.

   Implementations MAY support the "format" keyword as a validation
   assertion.  Should they choose to do so:

      they SHOULD implement validation for attributes defined below;

      they SHOULD offer an option to disable validation for this
      keyword.
```

further below formats for date and time are specified (https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-validation-01#section-7.3.1):
```
These attributes apply to string instances.

   Date and time format names are derived from [RFC 3339, section 5.6](https://datatracker.ietf.org/doc/html/rfc3339#section-5.6)
   [[RFC3339](https://datatracker.ietf.org/doc/html/rfc3339)].

   Implementations supporting formats SHOULD implement support for the
   following attributes:

   date-time:  A string instance is valid against this attribute if it
      is a valid representation according to the "date-time" production.

   date:  A string instance is valid against this attribute if it is a
      valid representation according to the "full-date" production.

   time:  A string instance is valid against this attribute if it is a
      valid representation according to the "full-time" production.
```

As you can see, the specification is not explicitly stating that a `format: date` MUST be validated as such, but it SHOULD be (it depends on the JSON schema implementation whether this happens or not). The current implementation is cautious and doesn't block submitting a form if the input doesn't match the specified format.

While it is possible there are connectors which currently rely on this behavior (stating a `format` but actually accepting different inputs as well), to me the risk of breaking these in a small way (if the user wants to edit an existing configuration the problem will be flagged and the user can fix the date format) does not outweigh the benefit of removing a common source of user errors for these fields. Based on this I suggest the following steps:
* Introduce a default regex pattern enforced in the frontend if `format` is set to `time` or `date-time`
* If a custom pattern is set via spec, the custom pattern is used instead of the default pattern
* Add a feature flag so this behavior can be rolled back if necessary

**Comment by flash1293** [comment]
What do you think @timroes @bnchrch @lmossman  ?

**Comment by josephkmh** [comment]
> Acceptance tests warn if the format and the pattern doesn't match (but it doesn't fail validation)

Is it a big lift to fail the acceptance tests if the format and pattern don't match? Or could we enforce that `pattern` must be null if `format` is populated, since we are treating that field as a proper sub-type? Otherwise I think there is still the gap that a connector developer might use `format: date-time` together with a non-standard `pattern`, which is handled poorly in the UI.

**Comment by timroes** [comment]
My two cents would be: We should forbid any pattern on a date(time) field via CAT and remove them where they still exist and make sure to build a validation for those types into the FE. I don't think having the spec define a pattern for date fields is a great solution, since in theory (unless we make the CAT tests a failure) can vary from our understanding how the date needs to look. Also the error messaging on missmatched patterns isn't really great, since we don't have a way for the spec to give us a nicer human readable error message if the pattern fails atm, so they'd just see: "input doesn't match <regex>" which isn't great. I feel handling this completely at the FE (and forbid connectors to specify a pattern for date fields), so we can ensure our validation is always in sync with our actual date picker component, would be the best option.

**Comment by bnchrch** [comment]
Well put Tim.

Let me just add a couple supportive points.

1. When it comes to datetime fields I imagine most follow ISO8601 and because of that I dont think its reasonable for the majority of our connectors to copy and paste the following regex to ensure validity.
```
/^([\+-]?\d{4}(?!\d{2}\b))((-?)((0[1-9]|1[0-2])(\3([12]\d|0[1-9]|3[01]))?|W([0-4]\d|5[0-2])(-?[1-7])?|(00[1-9]|0[1-9]\d|[12]\d{2}|3([0-5]\d|6[1-6])))([T\s]((([01]\d|2[0-3])((:?)[0-5]\d)?|24\:?00)([\.,]\d+(?!:))?)?(\17[0-5]\d([\.,]\d+)?)?([zZ]|([\+-])([01]\d|2[0-3]):?([0-5]\d)?)?)?)?$/
```

2. Because we already show a datepicker with `date` and `date-time` fields we essentially already are asking users use a specific date time format. We just dont currently validate in the case they deviate. 

The only reason we may want them to override format is they have a strange datetime format. However in that case i'd argue our datetime picker no longer serves a purpose and they should use a string field.

**Comment by flash1293** [comment]
I think we are all in agreement here, summarizing the changes:
* Introduce a default regex pattern enforced in the frontend if format is set to time or date-time (like defined in the JSON schema specification )
* Do not allow `pattern` along with the format
* Update acceptance tests to validate this
* Fix existing cases where pattern and format are specified together and audit all cases where `format` is used that it will work with the default formats produced by the datepicker
* Document this behavior

**Comment by lmossman** [comment]
I agree with Tim's suggestion and the above summary, I think standardizing our handling of datetimes by disallowing a `pattern` when `format` is set makes sense, and will make both connector developers' and airbyte users' lives easier.

And in the future, maybe we can extend the default regex patterns we are enforcing to include more formats like `email`, `uuid`, etc.
